### PR TITLE
feat(integration/fivetran): Fivetran source ingestion integration

### PR DESCRIPTION
**Pull request by shubhamjagtap639**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/setup.py [hunk 1]
```diff
@@ -505,6 +506,7 @@ def get_long_description():
             "nifi",
             "vertica",
             "mode",
+            "fivetran",
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/setup.py:529]: why is this needed?

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+SUPPORTED_DATA_PLATFORM_MAPPING = {
+    "postgres": "postgres",
+    "snowflake": "snowflake",
+    "mysql": "mysql",
+}
+
+
+class SnowflakeDestinationConfig(BaseSnowflakeConfig):
+    database: str = Field(
+        default=None, description="The fivetran connector log database."
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: Is there any default for the database?

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        default="FIVETRAN_LOG", description="The fivetran connector log schema."
+    )
+
+
+class FivetranLogConfig(ConfigModel):
+    destination_platform: str = pydantic.Field(
+        default=None,
+        description="The destination platform where fivetran connector log tables are dumped.",
+    )
+    snowflake_destination_config: Optional[SnowflakeDestinationConfig] = pydantic.Field(
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: let's rename it to `destination_config`

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        default=None, description="The fivetran connector log database."
+    )
+    log_schema: Optional[str] = Field(
+        default="FIVETRAN_LOG", description="The fivetran connector log schema."
+    )
+
+
+class FivetranLogConfig(ConfigModel):
+    destination_platform: str = pydantic.Field(
+        default=None,
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: Can we set it to default `snowflake` as we are currently supporting only `snowflake`? 
let's also add a validation for valid values of this argument.

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        description="Regex patterns for connectors to filter in ingestion.",
+    )
+    # Configuration for stateful ingestion
+    stateful_ingestion: Optional[StatefulStaleMetadataRemovalConfig] = pydantic.Field(
+        default=None, description="Airbyte Stateful Ingestion Config."
+    )
+    # Fivetran connector all sources to platform instance mapping
+    sources_to_platform_instance: Dict[str, PlatformDetail] = pydantic.Field(
+        default={},
+        description="A mapping of connector all sources to Data platform instance. Provide connector id as key.",
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: A mapping of the connector's all sources to platform instance. Use connector id as key.

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    )
+    # Fivetran destination to platform instance mapping
+    destination_to_platform_instance: Dict[str, PlatformDetail] = pydantic.Field(
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py:131]: Could you please add more documents to identify what is difference between this destination and destination_config?

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/fivetran.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+            result_type=Constant.ORCHESTRATOR,
+        ):
+            yield mcp.as_workunit()
+
+    def _get_connector_workunit(
+        self, connector: Connector
+    ) -> Iterable[MetadataWorkUnit]:
+        self.report.report_connectors_scanned()
+        # Create dataflow entity with same name as connector name,
+        # Later soft delete this entity during ingestion
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/fivetran.py]: Coul you please add information on why are we doing soft delete?

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/fivetran_schema.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
+import json
+import logging
+from dataclasses import dataclass
+from typing import List, Tuple
+
```
Comment (siddiquebagwan-gslab) [inline metadata-ingestion/src/datahub/ingestion/source/fivetran/fivetran_schema.py]: Could you please rename this file to dataclasses.py or domain_model.py

**Review by siddiquebagwan-gslab — CHANGES_REQUESTED** [review]
Could you please add a document like we do have in Power BI to mention the concept mapping

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py (hunk 2 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: No
Even for log_schema config there is no default hence removing default value

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py (hunk 3 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: But this config is specific to snowflake destination only.

File: metadata-ingestion/setup.py (hunk 1 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/setup.py:529]: Now we added one unit test case on fivetran source

File: metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py (hunk 3 above)
Comment (siddiquebagwan-gslab) [reply metadata-ingestion/src/datahub/ingestion/source/fivetran/config.py]: type is sufficient to understand `SnowflakeDestinationConfig`, later we can change it to Union[SnowflakeDestinationConfig|BigQueryDestinationConfig], so let's rename it to `destination_config`

**Review by siddiquebagwan-gslab — CHANGES_REQUESTED** [review]

**Comment by hsheth2** [comment]
@shubhamjagtap639 can we close this PR?

**Comment by shubhamjagtap639** [comment]
> @shubhamjagtap639 can we close this PR?

Yes, I am closing this PR
